row_cache.cc

```diff
@@ -637,12 +633,8 @@ class scanning_and_populating_reader final : public flat_mutation_reader_v2::imp
     }
 
     future<flat_mutation_reader_v2_opt> read_from_secondary() {
-        return _secondary_reader().then([this] (range_populating_reader::read_result&& res) {
-            auto&& [fropt, ps] = res;
+        return _secondary_reader().then([this] (flat_mutation_reader_v2_opt&& fropt) {
             if (fropt) {
-                if (ps) {
-                    push_mutation_fragment(std::move(*ps));
-                }
                 return make_ready_future<flat_mutation_reader_v2_opt>(std::move(fropt));
             } else {
                 _secondary_in_progress = false;
```

Comment:
> However, this makes a problem:
> now we pop partition_start from the underlying reader
> in autoupdating_underlying_reader::move_to_next_partition
> and manually push it back to downstream readers
> bypassing nonforwardable reader. This means if we
> change the logic in nonforwardable reader as described
> we will end up with partition_start without partition_end
> in the downstream readers.

I don't think make_nonforwardable() should rely on this assumption, that is that it is invoked on a stream which was never consumed from. It's enough if make_nonforwardable()::reader::next_partition() sets a flag which puts the reader at EOS without emitting anything.

Reply:
But the patch is good as a cleanup.

Reply:
> It's enough if make_nonforwardable()::reader::next_partition() sets a flag which puts the reader at EOS without emitting anything.

It's true for the case when `single_partition=true`, and this indeed should be fixed/tested. In general, there may be other partitions and we should handle that if `make_nonforwardable` pretends to be generic. In particular, we have to handle situations when this next partition is empty or not empty. In the first case we shouldn't emit `partition_end`, in the second case we should. The `_partition_is_open` flag was added to distinguish these two cases. Here we strive to preserve the semantics of intra-partition fast-forwarding mode described in [your comment](https://github.com/scylladb/scylladb/issues/12249#issuecomment-1439061442).